pkg/cluster/kubeclient/roundtrip.go

```diff
@@ -1,161 +1,88 @@
-//    timeout.
-//
-// 3. Even with the default 30 second Dial timeout, sometimes we see unexplained
-//    Dial timeout failures.  RetryingRoundTripper Retries in these cases.
-//
-// 4. The default TLS handshake timeout is 10 seconds.  Sometimes we see this
-//    timeout triggered.  RetryingRoundTripper also Retries in these cases.
-
-var timerExpired = errors.New("RetryingRoundTripper timer expired")
-
-type RetryingRoundTripper struct {
-	Log *logrus.Entry
-	http.RoundTripper
-	Retries    int
-	GetTimeout time.Duration
-}
-
-func (rt *RetryingRoundTripper) RoundTrip(req *http.Request) (resp *http.Response, err error) {
-	var retry int
-	for {
-		retry++
-
-		if req.Method == http.MethodGet {
-			done := make(chan struct{})
-
-			cancel := make(chan struct{})
-			req.Cancel = cancel
-
-			t := time.NewTimer(rt.GetTimeout)
-
-			go func() {
-				select {
-				case <-done:
-				case <-t.C:
-					close(cancel)
-				}
-			}()
-
-			resp, err = rt.RoundTripper.RoundTrip(req)
-
-			if !t.Stop() {
-				err = timerExpired
-			}
-
-			close(done)
-
-		} else {
-			resp, err = rt.RoundTripper.RoundTrip(req)
-		}
-
-		if err, ok := err.(*net.OpError); retry <= rt.Retries && ok {
-			// grr, "i/o timeout" is defined in internal/poll/fd.go and is thus
-			// inaccessible
-			if err.Op == "dial" && err.Err.Error() == "i/o timeout" {
-				rt.Log.Warnf("%s: retry %d", err, retry)
-				continue
-			}
-		}
-
-		// TODO: on the few occasions I've seen this, it's been down to an API
-		// server crash.  Need to investigate further.
-		if retry <= rt.Retries && utilerrors.IsMatchingSyscallError(err, syscall.ECONNREFUSED) {
-			rt.Log.Warnf("%s: retry %d", err, retry)
-			continue
-		}
-
-		// grr, http.tlsHandshakeTimeoutError is not exported.
-		if retry <= rt.Retries && err != nil && err.Error() == "net/http: TLS handshake timeout" {
-			rt.Log.Warnf("%s: retry %d", err, retry)
-			continue
-		}
-
-		if retry <= rt.Retries && err == timerExpired {
-			rt.Log.Warnf("%s: retry %d", err, retry)
-			continue
-		}
+// NewKubeclient creates a new kubeclient
+// If PrivateEndpointIp is not nil - kubeclient roundtripper will point PrivateEndpoint IP address
+func NewKubeclient(log *logrus.Entry, config *v1.Config, cs *api.OpenShiftManagedCluster, disableKeepAlives bool, testConfig api.TestConfig) (Interface, error) {
+	restconfig, err := NewRestConfig(log, config, cs, disableKeepAlives, testConfig)
+	if err != nil {
+		return nil, err
+	}
 
+	if cs.Properties.NetworkProfile.PrivateEndpoint == nil {
+		restconfig.WrapTransport = roundtrippers.NewRetryingRoundTripper(log, disableKeepAlives)
+	} else {
+		tlsConfig, err := restclient.TLSConfigFor(restconfig)
 		if err != nil {
-			rt.Log.Warnf("%#v: not retrying", err)
+			return nil, err
 		}
 
-		return
+		restconfig.WrapTransport = roundtrippers.NewPrivateEndpoint(log, cs, disableKeepAlives, testConfig, tlsConfig)
+		if testConfig.RunningUnderTest {
```

Comment:
This is partially my code base, but reading it with fresh eyes - we need much better comment here.